internal/uvm/virtual_device.go

```diff
@@ -0,0 +1,39 @@
+package uvm
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/Microsoft/hcsshim/internal/requesttype"
+	hcsschema "github.com/Microsoft/hcsshim/internal/schema2"
+)
+
+const (
+	resourcePathVpciFmt = "VirtualMachine/Devices/VirtualPci/%s"
+)
+
+func (uvm *UtilityVM) GpuCounter() uint64 {
+	return atomic.AddUint64(&uvm.gpuCounter, 1)
+}
+
+func (uvm *UtilityVM) AssignDevice(ctx context.Context, id string, device hcsschema.VirtualPciDevice) error {
```

Comment:
The _ethos_ of the other functions is to try and not expose the HCS schema outside of the UVM layer. Is there a reason we cant simply pass AssignDevice(ctx, instancePath string, vf *uint) (id, error) "id" is can be completely managed in this function. And if you dont want to manage a separate ID we know that instancePath has to be unique so it can act as the ID already.

Reply:
I don't see an issue with doing that. I'll still need to manage a separate id though. nvidia-container-cli expects either a device index or device uuid. I'm not aware of any way for us to figure out the device uuid in the code, so instead I'm using the gpuCounter to represent the gpu devices on the uvm so we know the device index :)

Reply:
Probably for this to be reliable we are going to have to find a way to map from the host identifier to something in the /proc/bus/pci namespace in the guest, to the nvidia GPU index or UUID.

Reply:
We might have to ask the HCS team to allow us to configure the VMBus VPCI channel instance ID so that we can accurately perform this mapping. I'm not sure.

Reply:
What do you mean by "configure the VMBus VPCI channel instance ID"?

Reply:
I still dont get this code. How does the index correlate from host to guest? As Starks said it seems like we would need the UUID or something.

Reply:
That's what I'm working on right now. We can pass a GUID through as the device ID to HCS, which will then use that GUID for the VMBUS path. But that seems to be the only thing we have control of. From the VMBUS path, we can find the bus location (under some path like /sys/vmbus/devices/<GUID>/...) of the GPU and then with modifications to nvidia-container-cli, we can pass that through to uniquely identify the device instead of the gpu UUID or index.

Reply:
Does that make sense?

Reply:
Yep. I like it